### ISSUE TITLE
Disable anti-alias on font rendering

### ISSUE THREAD

**Issue by melroy89** [raphamorim/rio#732]
Me again. 😆  Since Rio is rendering from it's own GPU engine it will not use the `~/.config/fontconfig/fonts.conf` like all my other programs.

For example I use:

```sh
  <match target="pattern">
    <test qual="any" name="family"><string>DejaVuSansM Nerd Font Mono</string></test>
    <edit name="antialias" mode="assign"><bool>false</bool></edit>
    <edit name="hintstyle" mode="assign"><const>hintslight</const></edit>
    <edit name="hinting" mode="assign"><bool>false</bool></edit>
    <edit name="rgba" mode="assign"><const>rgb</const></edit>
    <edit name="lcdfilter" mode="assign"><const>lcddefault</const></edit>
  </match>
```

Yeah, I really dislike anti-alias. I use a very specific font that is scrip (especially when) anti-alias is **disabled**. Which looks so much better (believe me).

It would therefor be nice to have a rendering option to turn off anti-alias.

---

For reference, here is a next to next comparison (left is my old XFCE terminal, anti-alias off, hinting off. On the right is Rio).

![image](https://github.com/user-attachments/assets/c9c0c674-5458-4a03-bffb-a93aa88da804)

(**Important:** Open the image is a new tab, so you don't get browser blurry artifacts!)

**Comment by raphamorim** [comment]
Implemented part of it on [5d6bb98](https://github.com/raphamorim/rio/commit/5d6bb989a6937e240c577f5e6d70ce372eea3cba) (now can disable hinting)

**Comment by melroy89** [comment]
> Implemented part of it on [5d6bb98](https://github.com/raphamorim/rio/commit/5d6bb989a6937e240c577f5e6d70ce372eea3cba) (now can disable hinting)

You're just as quick as me ;P well done with the hinting option already!

**Comment by melroy89** [comment]
OK, debugging & learning the project for about 2 hours now (plus I'm new to Rust.. 😆 ).

_TLDR:_ I have no idea to disable anti-alias, haha!

Correct me if I'm wrong, but here is my story:

**Background story:**

You first load all the fonts/glyphs in the font loader. Sugarloaf gives you a content, in the content you can add for example text (`add_text`). Then your text is added to `fragments` with an fragment style. Then during a `build` you convert all the lines and other stuff and move it into `render_data`. Then somehow, a bit too much code for me now.. eventually... during rendering I guess, it will use the `rich_ext` module with the `RichTextBrush`. 

This is using a `wgpu::SamplerDescriptor`, which is this struct in the wgpu crate: https://docs.rs/wgpu/latest/wgpu/struct.SamplerDescriptor.html

And I guess this wgpu struct is responsible for rendering the text using the glyphs? And looking at this:

https://github.com/raphamorim/rio/blob/main/sugarloaf/src/components/rich_text/mod.rs#L131

The `wgpu::SamplerDescriptor` has already the following options `nearest`, but there is no option `None`:

```rs
            address_mode_u: wgpu::AddressMode::ClampToEdge,
            address_mode_v: wgpu::AddressMode::ClampToEdge,
            address_mode_w: wgpu::AddressMode::ClampToEdge,
            mag_filter: wgpu::FilterMode::Nearest, // <-- You can't set it to None or something
            min_filter: wgpu::FilterMode::Nearest, // <-- You can't set it to None or something
            mipmap_filter: wgpu::FilterMode::Nearest ,// <-- You can't set it to None or something
            lod_min_clamp: 0f32,
            lod_max_clamp: 0f32,
```

Long story, short I'm not sure if wgpu can even render without anti-alias? Maybe it's just me...

Again, also if I got something wrong (and most likely missing in my story). Let me know 😄 , like I said I'm also still learning this code base. After all rio is "only" 186186 lines of code.

**Comment by raphamorim** [comment]
> OK, debugging & learning the project for about 2 hours now (plus I'm new to Rust.. 😆 ).
>
> TLDR: I have no idea to disable anti-alias, haha!
>
> Correct me if I'm wrong, but here is my story:

haha love it!

The missing piece of turn anti-alias off (considering that hinting is off as well) is disable subpixel rendering and use https://docs.rs/zeno/latest/zeno/enum.Format.html#variant.Alpha (if i am not mistaken)

https://github.com/raphamorim/rio/blob/b53acf9c37428e87fdca1b537567c07063ca0965/sugarloaf/src/components/rich_text/image_cache/glyph.rs#L150-L152

**Comment by melroy89** [comment]
OK check, it seems `CustomSubpixel` is the one indeed. Since a `format()` is required to set to fill the glyphs, I just used:

```rust
.format(Format::CustomSubpixel([0., 0., 0.]))
```

That seems to disable anti-alias.

However, this won't look still not very good under the `text` example. I tried.. Since I dunno what you're doing there, but setting the `font_size` and `sugarloaf_layout` it not really making it better. 
https://github.com/raphamorim/rio/blob/b53acf9c37428e87fdca1b537567c07063ca0965/sugarloaf/examples/text.rs#L57

So I also tried the `multi_text` example, and the text in the right box, looks the best thus far for sure (maybe it has no scale factor?).. Anyhow, look at this:

Before (subpixel rendering enabled):

![image](https://github.com/user-attachments/assets/6760dd0c-0d90-4890-82a3-f8574c42a48e)

After (subpixel rendering disabled, aka disabled anti-alias):

![image](https://github.com/user-attachments/assets/ca799c6c-7dac-44ba-818b-b0f9c68708e9)

Or gif (1:1 ratio):

![before_after_subpixel_rendering](https://github.com/user-attachments/assets/28df1ae0-5746-4458-861f-1fad83b97ce3)

As you can see disabling subpixel rendering make the font so much sharper. 

Or this video (zoomed-in):

https://github.com/user-attachments/assets/7b56399d-3249-49ee-930a-bbf4abc97cbb

But just note that increasing the font-size (scaling or whatever) can and most likely make the text blurry again:

![image](https://github.com/user-attachments/assets/d5fab2eb-07a2-41b3-a169-7610efcf85f0)


And a comparison between XFCE terminal with Rio terminal. Despite disabling supixel rendering, the XFCE terminal still seem to produce sharper text somehow. Yes, I use the same font. Yes, I also disabled hinting in Rio settings...

![image](https://github.com/user-attachments/assets/3700e359-80f4-4e74-9ff3-221d1da0d9a0)

**Comment by Null-Kelvin** [comment]
Any updates?

**Comment by raphamorim** [comment]
partial progress in #1715: fonts.hinting now applies to the terminal grid path too (it was hardcoded on). the anti aliasing / subpixel / fontconfig part is still open
